File: chainofresponsibilitypattern.py
from enum import Enum

class LogLevel(Enum):
    INFO = 1
    DEBUG = 2
    ERROR = 3
    
class LogProcessor:
    def __init__(self, logger_processor):
        self.next_logger_processor = logger_processor

    def log(self, log_level, message):
        if self.next_logger_processor:
            self.next_logger_processor.log(log_level, message)

class InfoLogProcessor(LogProcessor):
    def __init__(self, next_logger_processor):
        super().__init__(next_logger_processor)

    def log(self, log_level, message):
        if log_level == LogLevel.INFO:
            print('INFO: ', message)
        else:
            super().log(log_level, message)

class DebugLogProcessor(LogProcessor):
    def __init__(self, next_logger_processor):
        super().__init__(next_logger_processor)

    def log(self, log_level, message):
        if log_level == LogLevel.DEBUG:
            print('DEBUG: ', message)
        else:
            super().log(log_level, message)

class ErrorLogProcessor(LogProcessor):
    def __init__(self, next_logger_processor):
        super().__init__(next_logger_processor)

    def log(self, log_level, message):
        if log_level == LogLevel.ERROR:
            print('ERROR: ', message)
        else:
            super().log(log_level, message)

if __name__ == '__main__':
    logObject = InfoLogProcessor(DebugLogProcessor(ErrorLogProcessor(None)))

    logObject.log(LogLevel.ERROR, "exception happens")
    logObject.log(LogLevel.DEBUG, "need to debug this ")
    logObject.log(LogLevel.INFO, "just for info ")

File: chainofresponsibilitypattern/chainofresponsibilitypattern.js
// ===== Enum for Log Levels =====
const LogLevel = Object.freeze({
  INFO: 1,
  DEBUG: 2,
  ERROR: 3,
});

// ===== Abstract Log Processor =====
class LogProcessor {
  constructor(nextLoggerProcessor) {
    this.nextLoggerProcessor = nextLoggerProcessor;
  }

  log(logLevel, message) {
    if (this.nextLoggerProcessor) {
      this.nextLoggerProcessor.log(logLevel, message);
    }
  }
}

// ===== Concrete Info Logger =====
class InfoLogProcessor extends LogProcessor {
  constructor(nextLoggerProcessor) {
    super(nextLoggerProcessor);
  }

  log(logLevel, message) {
    if (logLevel === LogLevel.INFO) {
      console.log('INFO: ', message);
    } else {
      super.log(logLevel, message);
    }
  }
}

// ===== Concrete Debug Logger =====
class DebugLogProcessor extends LogProcessor {
  constructor(nextLoggerProcessor) {
    super(nextLoggerProcessor);
  }

  log(logLevel, message) {
    if (logLevel === LogLevel.DEBUG) {
      console.log('DEBUG: ', message);
    } else {
      super.log(logLevel, message);
    }
  }
}

// ===== Concrete Error Logger =====
class ErrorLogProcessor extends LogProcessor {
  constructor(nextLoggerProcessor) {
    super(nextLoggerProcessor);
  }

  log(logLevel, message) {
    if (logLevel === LogLevel.ERROR) {
      console.log('ERROR: ', message);
    } else {
      super.log(logLevel, message);
    }
  }
}

// ===== Client Code =====
const logObject = new InfoLogProcessor(
  new DebugLogProcessor(
    new ErrorLogProcessor(null)
  )
);

logObject.log(LogLevel.ERROR, "exception happens");
// ERROR:  exception happens

logObject.log(LogLevel.DEBUG, "need to debug this");
// DEBUG:  need to debug this

logObject.log(LogLevel.INFO, "just for info");
// INFO:  just for info
